src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs

```diff
@@ -76,9 +75,10 @@ internal sealed class ReflectionEmitMemberAccessor : MemberAccessor
             ParameterInfo[] parameters = constructor.GetParameters();
             int parameterCount = parameters.Length;
 
-            if (parameterCount > JsonConstants.MaxParameterCount)
+            if (parameterCount > 64)
             {
-                return null;
+                // Avoid emitting dynamic code for constructors with large parameter counts
```

Comment:
This fallback is still needed because large enough dynamic methods appear to be failing with `InvalidProgramException`. 

Not sure why, I couldn't find any documentation on size restrictions cc @dotnet/area-system-reflection-emit

Reply:
It might be the case that IPE's being thrown here might have motivated the initial constructor parameter limit being introduced.

Reply:
What are the argument counts that you are seeing failures with?

If you are seeing `InvalidProgramException` with say less than 1000 arguments (that are not huge structs), it likely means that there is a bug in your ref emit code.

Reply:
This `Ldc_I4_S` does not look right:

https://github.com/dotnet/runtime/pull/75982/files/897ea4a024d88c0e56b666165e39015e18bd51ed#diff-c0e6bc47405dcd7e6920cb0f76f369ef6a99a7ffca192e9cc581dbcfb6e01c8eR98

`Ldc_I4_S` is only able to load integer that is <128.

Reply:
> If you are seeing InvalidProgramException with say less than 1000 arguments

Seeing it with exactly 1000 arguments (cf. Newly added test in this PR)

Reply:
1000 int arguments should not be a problem. It is likely the bug that I have highlighted.

Reply:
Changing the opcode fixed the issue. Still, I suspect there might be diminishing returns in emitting a dynamic method after a large enough number of constructor parameters. Do you think some threshold should be kept?

Reply:
> Still, I suspect there might be diminishing returns in emitting a dynamic method after a large enough number of constructor parameters. Do you think some threshold should be kept?

What does benchmarking show?

Reply:
I do not think that this needs any thresholds.

Note that the constructor receiving the arguments has to be JITed too. If you cross internal implementation limits for large number of arguments in codegen, you are likely the hit the problem when JITing the constructor anyway. Adding a thresholds here would just delay the problem a bit, but not lead to success (except maybe some corner cases).

Reply:
Here's deserializing 1000 constructor arguments. Performance is "same" but allocation profile is increased, presumably because the reflection member accessor needs to copy the argument array due to [inexact array lengths because of buffer pooling](https://github.com/dotnet/runtime/blob/161464e33a1333bb5ac3fc509172a5e2ada5af04/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionMemberAccessor.cs#L60-L62):

|                   Method |       Mode  |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio | MannWhitney(3%) | RatioSD |  Gen 0 | Allocated | Alloc Ratio |
|------------------------- |----------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|---------------- |--------:|-------:|----------:|------------:|
| DeserializeFromUtf8Bytes |  Reflection | 241.1 us | 11.55 us | 13.30 us | 240.0 us | 220.7 us | 272.4 us |  1.00 |            Base |    0.00 | 6.6288 |  73.43 KB |        1.00 |
| DeserializeFromUtf8Bytes | Reflection.Emit | 229.9 us | 10.14 us | 11.27 us | 226.7 us | 212.2 us | 252.4 us |  0.96 |            Same |    0.06 | 5.3619 |  57.76 KB |        0.79 |

Reply:
Seeing hard crashes in the interpreted mono test runs. I'm guessing large dynamic methods is causing the mono interpreter to crash. I'm actually surprised we're using Ref.Emit in the interpreted runtimes:

https://github.com/dotnet/runtime/blob/9b9b0c95c678065044d01049c56915e6b750b10a/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L598-L601

What is our guidance concerning use of dynamic code in the interpreter/wasm? One possibility is disabling it altogether using `RuntimeFeature.IsDynamicCodeCompiled` here instead. cc @eerhardt

Reply:
Do the crashes with large number of constructor arguments go away when you use reflection-based path?

> I'm guessing large dynamic methods are causing the mono interpreter to crash. 

The interpreter works pretty much the same for both dynamic and non-dynamic methods. If you are seeing large dynamic method to crash the interpreter, it is because the method is large. It is not because the method is dynamic. You would see the same crash with equivalent non-dynamic method too.

Reply:
> I'm actually surprised we're using Ref.Emit in the interpreted runtimes:

I know these numbers are quite old by now, but in the .NET 6 timeframe we measured and found that using Ref.Emit in the WASM runtime was faster for both serialize and deserialize. See https://github.com/dotnet/runtime/issues/38693#issuecomment-842704244. That's why we've never switched to using `IsDynamicCodeCompiled` so far - we would degrade throughput of basic serialization scenarios.

Reply:
> we measured and found that using Ref.Emit in the WASM runtime was faster for both serialize and deserialize.

If the findings were specific to STJ, it could be related to the fact that the reflection accessors need to copy argument arrays: https://github.com/dotnet/runtime/pull/75982#discussion_r977670487. There might be performance benefits if we explicitly disable buffer pooling for object arrays when using the reflection member accessor.

Reply:
> Do the crashes with large number of constructor arguments go away when you use reflection-based path?

It appears that the crashes persist, even when reflection is used. So @jkotas is right, the 1000 parameter test class must be hitting limits inherent in the interpreter. I'll simply disable the test for the interpreter and revert the fallback commit.

Reply:
cc @BrzVlad Mono interpreter crashing on methods with 1000 arguments.

Reply:
@eiriktsarpalis The 1000 argument limit should be arbitrary in the interpreter, probably a remnant from its prototype. I will soon merge https://github.com/dotnet/runtime/pull/76092. If you want you could rerun your PR afterwards, or I could also reenable the test on my PR if this is ready to go already.

Reply:
@BrzVlad feel free to merge, will rebase my changes on Monday.

Reply:
Can confirm that the test no longer crashes interpreted test legs.